### ISSUE TITLE
Generate Word collection documents

### ISSUE THREAD

**Issue by opoudjis** [metanorma/metanorma#363]
Until now, we have ignored Word in Collection outputs. But there is a need to generate Word output as a single document, that has arisen in https://github.com/metanorma/metanorma-itu/issues/514

What that means is passing the Presentation XML for an entire collection to Word for rendering. Of course, isodoc/word does not understand collections, so that Presentation XML needs to be reduced to the Presentation XML of a single document. But that single document will not have the structure of a normal single document. It may have interleaved title sections; it may have interleaved prefatory sections; it may have stop-start page numbering. We may ignore the bibliographic metadata of each component document, or we may use it to generate a title page for each included document. 

This ticket is going to be somewhat trial and error, but it is needed, to realise the ITU requirement of a contribution document, embedding a proposed document — appearing as an annex, with consecutive page numbering, but with a discrete title page. And we will gradually genericise what kinds of document model the metanorma collection will support in Word.

**Comment by opoudjis** [comment]
From that ticket, this is the motivating typology of document rendering for this ticket:

We have had a similar instance to this in JIS commentaries, which are included in the same Word document as a standard but paginated differently, and with a different header.

But we need to formalise this in Metanorma for Word documents specifically, because this is going to keep coming up, and we need a coherent approach rather than ad hoc solutions (such as we arguably did for JIS).

In Metanorma, we have a concept of a **document**: something that follows the Metanorma document schema, with a preface, clauses, and annexes in that order. The clauses are numbered consecutively from beginning to end. There is a single cover page, a single instance of introductory boilerplate, and a single table of contents.

We have instances where what were originally separate documents are combined into a single document, and are processed as a single document. That is **embedded documents**. So whatever the numbering of the clauses in the source document, the resulting artefact is a single document, with a single numbering of clauses from beginning to end, and a single cover and intro page. The way they have been implemented to date, the embedded document is the main set of clauses, and any wrapping content is additional prefatory content (or potentially additional annexes). That's how there can still be a single numbering of clauses: there is a single document, with stuff added before and after it. 

We also have had instances where multiple documents reference each other, and are compiled as a **document collection**, dynamically updating those cross-references as hyperlinks. These are treated as a document collection: they end up all being aggregated into a single XML artefact, and the cross-references are preprocessed across the entire collection, so that they render consistently as hyperlinks. But each individual document in the collection is processed as a standalone document: the only thing that links the documents together is the hyperlinks and an overall cover page and index.

Document collections to date have been implemented as HTML and PDF, not Word. 

We have implemented the following to date:

* ITU Recommendation Annexes. To date, we have only ever processed these as standalone documents. We have never attempted to render them alongside their original document, and if we had, I think we would have run into this issue a lot earlier.
* ISO 10303. The various parts of this standard, e.g. ISO 10303-41, are standalone documents, published separately in HTML and PDF, with their own coverpage and clause numbering. But documents in this standard hyperlink each other, and are published as a web site. So they are realised as a document collection.
* BIPM SI Brochure. This contains two separate documents with the same content, in English and in French. Each document has its own title and intro page, and its own clause numbering. There is one instance in which they hyperlink each other. They too are realised as a document collection.
* BSI national prefaces, wrapping CEN and ISO documents. They are additional national prefaces and a BSI coverpage, prefixed to a CEN or ISO document; the CEN document may itself be a regional prefix attached to an ISO document. The result is a single document, with a single (BSI) cover page, and a single (ISO) clause numbering. This is what we created embedded documents for, and they are indeed primarily intended for national adaptions of international/regional standards—which do not alter the main body of the standard, nor do they present the international standard as a distinct document.
* JIS Commentaries. These  have a distinct Word header, and distinct pagination, but in itself that is not enough to treat them as distinct documents: they are numbered as annexes to the main document. They could have been treated as embedded documents, but because they are annexes, and aren't circulated as discrete documents, I have not yet bothered to do even that much.

That brings us to ITU contributions. On the face of it, they are Embedded documents. They add prefatory material to an separately circulated output document, presenting it as a single document. They maintain the same document header, and consecutive page numbering. 

But:

* Revised documents and New Work Items both retain a separate cover page and Table of Contents for the embedded document.
* New Work Items involve Annexes (Form A.1/A.13, and the New Work Item itself): so the embedding document does have different clause  numbering to the embedded document. (If Metanorma processes these as a single document, it can't jump back from Annex 1 back to a document preface).
* The Work Item claims to be an Annex, but it has a preface, a title page, and Appendixes, and it is numbered as the source document, not as an annex at all. Which means that calling it an annex has no impact on how it is rendered: it is rendered as a standalone document.

I can tweak the embedded solution to cope with these requirements—automatically extracting a separate cover page somehow, and inserting a table of contents into the embedded document. But the embedding happens as an Asciidoctor include, during document preprocessing. The metadata of the embedded document is preserved in the metadata of the embedding document, as metadata about what the document is based on. But as far as Metanorma processing is concerned, this really is a single document; and telling Metanorma to jump back from annex to preface is going to break its processing model.

That's why the real solution to rendering these has to be a document collection. @ronaldtse is going to be unhappy about that, because document collections are more complicated to set up. I have no sympathy for that objection: document collections are going to be preconfigured in the mn-templates-itu, and I'm not going to destroy the Metanorma processing model to force what is clearly two separate documents to be processed as a single document, but end up looking like two separate documents anyway. These are two separate documents, and Metanorma needs to render one document at a time.

This is going to require added a lot of implementation effort. First, we have never dealt with document collections as  a single Word document at all—although making that happen will not be prohibitive. Second, we need directives on what the cover page and intro page of each document looks like in Word, which will be a truncation of the standalone cover and intro pages, passed as custom templates by the collection. 

Each Word document is structured as Title section (no page numbers), Preface section (usually Roman number pagination), Main section (Arabic number pagination). The workflow is going to be:

* Title + Preface + Main section of first document
* For each remaining document
  * Custom title document template, if supplied
  * Preface + Main section of that document

By default, each section (preface and main section of each document) will restart page numbering from 1, but I can introduce configuration to continue numbering.


In the case of ITU New Work Items, the custom title document will be:

![image](https://github.com/metanorma/metanorma-itu/assets/1911039/b1361714-e0d2-4531-b5af-d58c9b50c346)


````
Annex B

{{docidentifier}}

{{doctitle}}

Summary: {{abstract}}

Keywords: {{keywords}}
```` 

In other words, I am not proposing to mark up the embedded document as an annex at all, but to keep the annex as an artefact of the cover page. As I said above, the embedded document is not rendered as a true annex: it really is just treated as a separate document. So I'm not feeling guilty about that kind of subterfuge.

**Comment by opoudjis** [comment]
### Option 1: flatten collection to a single document

We have 

* metanorma-collection
  * bibdata0
  * manifest
  * prefatory-content
  * doc-container
    * standard-document1 (or flavour-equivalent, e.g. iso-standard)
      * bibdata1
      * preface1
      * sections1
      * annex1
      * bibliography 1 
  * doc-container
    * standard-document2 (or flavour-equivalent, e.g. iso-standard)
      * bibdata2
      * preface2
      * sections2
      * annex2
      * bibliography2
  * final-content

 We want to squash this into a single standard-document. We will start simple:

* standard-document
    * bibdata0
    * prefatory-content
    * preface1
    * sections1
    * annex1
    * bibliography 1 
    * preface2
    * sections2
    * annex2
    * bibliography2
    * final-content

Later on, we will work in bibdata1, bibdata2 as title pages, although I don't yet know how, or how flexibly to do so. We may not end up using the flavour title pages as templates; ITU contributions have a severely truncated title page for their embedded documents.

We will rely on the fact that isodoc can now parse clauses in whatever order we give it, using the displayorder attribute. The catch is that Word isodoc relies on the document structure to know when to initiate a new Section—title, preface, main, colophon, each with their own pagination; and that structure is now going to be broken. There is no longer going to be a single preface container.

In the  case of ITU contributions, we can ignore sections completely, since there is only one pagination and no title page. In the general case, we will need to refactor isodoc, so that sections are inserted as milestones—the transition from //preface to //sections, and now the transition from //sections or //annex back to //preface, or some other milestone indication that this is a new document, requiring a new Word section. We need to solve this for New Work Item proposals, which jump from Annex back to the Preface of the embedded document. 

That ends up becoming a flexible `make_body()` for Word: instead of fixedly iterating though title, preface, main, colophon (`make_body[1234]`), we need to identify the document bits and generate the sections as we go....

**Comment by opoudjis** [comment]
### Option 2: parse collections within isodoc/word

... Option 1 is increasingly looking worse to me, and it's worth considering parsing the collection in isodoc/word

* Generate a document shell `body`, as for single document
* body Section1: bibdata0
* For each doc-container:
  * body Section2: preface
  * body Section3: main
  * body Section3: footnotes, comments 

The doc-container parsing is just the existing parsing of a single document, and the rest of the flow happens as normal. Much less if any need to change existing single-document code, which Option 1 would force.

I'll give it a go...

**Comment by opoudjis** [comment]
After some shoving, including singleton methods defined on the render, to override population of the document shell with supplied XML (the Word HTML of the component documents), I have this working.

Next refinement: skip the title page and intro page generation for the embedded documents. We will want, more generally, to specify alternative Liquid template for the title and intro pages of embedded documents in Word, and a possible value of those templates will be nil. So:

````
directives:
  - documents-inline
  - word-titlepage: nil
  - word-intropage: nil
````

The next stage after that is to pass in a template for the truncated title page which ITU uses for its embedded documents in contributions.

Then, provide another directive, to ensure that pagination is continuous through the embedded sections, and does not restart at 1 for each document. That will involve either:

*  parsing the CSS to find the relevant `@page` definitions and postprocess them (which may not be a good idea, since Word does not use standard CSS)
* or embedding Liquid directives and alternate @page` definitions in every single flavour's Word CSS definitions

By the time that's done, we should have enough to realise ITU contributions.

**Comment by opoudjis** [comment]
Well... I've stripped the coverpage successfully, but the boilerplate is still being generated for internal documents, as is some internal title of other. I've lost a lot of sleep dealing with singleton methods last night, so I'm not continuing, but making good progress.

**Comment by opoudjis** [comment]
Getting errors with Nokogiri processing of raw nsbp in identifiers in ITU output passed back into Presentation XML; need to use HTMLEntities encoding more.

**Comment by opoudjis** [comment]
The test document G.650.1-201803 has a formula that is breaking Word: 

````
Loss(dB) = 10 log_(10)
{:( (P_{str})/(P_(bend))):}
````

Reporting it.

**Comment by opoudjis** [comment]
Now that I have the embedded document working again:

Something has changed in the rendering of ITU recommendations: content from WordSection2 is incorrectly showing up in WordSection1, and the cover page is mangled...

**Comment by opoudjis** [comment]
We need to specify the title page and intro page of the overall collection, and of individual included documents. Right now, we are only setting one title page and one intro page for the entire document, and we postprocess the entire document once. That will not work: we need to postprocess each document separately, with its own cover page potentially, and put them into the HTML file before we send it to html2doc for MIME bundling.

**Comment by opoudjis** [comment]
Need to refactor Word postprocessing, to isolate:

* Postprocessing of the document itself, cleanup() for generic postprocessing, word_cleanup() for Word-specific postprocessing
  * That includes the cover pages 
* Generation of the header and CSS
* Bundling of the Word document

The Metanorma collection Doc accesses the first of these when it generates individual documents.

**Comment by opoudjis** [comment]
Need to reset the Word converter in Metanorma for each document we process: we set the coverpage withoin the converter to nil for the first document, the contribution, but we don't want that change to persist for the next document, the embedded recommendation.

**Comment by opoudjis** [comment]
We now have separate cover/intro pages for the embedded and the overall document. So now the directives need to differentiate between the two.

We are letting the same CSS and header apply for the whole document, for now. This is already a time sink, and we can always come back to those later.

**Comment by opoudjis** [comment]
So now we have a document with all coverpages by default, we will tell it to turn all of them off:

````
directives:
  - documents-inline
  - collection-word-titlepage: null
  - collection-word-intropage: null
  - document-word-titlepage: null
  - document-word-intropage: null
````

And then we will turn just the document-word-titlepage on with a custom instance. Then, I'll call it a day...

**Comment by opoudjis** [comment]
Pretty good so far, but we have ongoing problems in confusion between postprocessing of the embedded documents. The abstract of the contribution, document 1, has ended up moving to the title page of the recommendation, document 2. 

The tension is, we want to do some postprocessing of Word output (certainly the cover pages), but not most of it (e.g. rearranging content—that should all be happening within the bounds of a single document.)

But we will have content in the embedding document we still want postprocessed. So we need a generic solution, without forcing flavour-specific code changes.

Simplest solution is: create the embedding document, with a placeholder to insert embedded content in. Postprocess that document, and then insert the already postprocessed embedded content, before sending it for rendering.

**Comment by opoudjis** [comment]
If there is no coverpage template in the current run, there must not be a section break with a next page, either.

**Comment by opoudjis** [comment]
We also need to delete the preface of a document, and its following section break, if it turns out to be missing. The preface to the ITU collection is.

**Comment by opoudjis** [comment]
Done. I still need to do the custom cover page, but I do also need to catch up with other tasks, so I'll pause this work here: the bulk of it is done (and it is forcing some refactoring on isodoc.)

**Comment by opoudjis** [comment]
Now that we are reparsing Word HTML > text output after postprocessing, in order to potentially add content to it, we need to add in namespaces, else the parse breaks.

**Comment by opoudjis** [comment]
New and surprising issue: the cover page for the embedded recommendation is missing its images in the Word file. The embedded recommendation itself isn't. So something is misconfiguring.

I don't need to solve this in general, since for embedded recommendations we will be skipping the cover page. But this is a new bug, and I do need to track it down. Then, custom embedded recommendation cover page template, then close this and set up ITU template for embedded recommendations.

**Comment by opoudjis** [comment]
It was a consequence of the fact that we are now processing namespaces in the Word doc properly: the element name of the `<v:imagedata` element that contains frames with images, as is typical of cover pages, is now `imagedata`, not `v:imagedata`, because the v: is being parsed.

**Comment by opoudjis** [comment]
Null value for coverpage is stil getting overwritten with flavour default, by design. We will introduce empty string for coverrpage instead in the config options, to force it not to be rendered at all.

**Comment by opoudjis** [comment]
We are now eliminating the embedded coverpage; as a result, there is now a redundant blank page, and a table of contents header but not table of contents. 

The table of contents is inserted into the `div[@class = 'toc']` placeholder inserted by the intro page template. if that template is skipped, the Table of Contents heading needs to be skipped too: the location of the table of contents is explicitly set by the intro page template.

This means that we need an embedded document intro page as well, consisting of only the table of contents placeholder div.

**Comment by opoudjis** [comment]
Accomplished what I need, from now on it's ITU-specific templating.

**Comment by opoudjis** [comment]
Writing tests, documentation

**Comment by opoudjis** [comment]
We fixed document identifiers within collections as keys of document Hash, to normalise non-breaking space to normal space. As a result, we've debugged the redundant addition of documents as both their bibdata (with space) and full documents (nonbreaking space); only the latter should be turning up, and overwriting the former which was intended as a placeholder.
